File: chatbot/backend/src/llm/interface.ts
import { ChatMessage, LLMConfig } from '../config'

/**
 * LLMProvider — Adapter interface for any LLM provider.
 * Each provider implements chat streaming and returns a ReadableStream of text chunks.
 */
export interface LLMProvider {
  /** Provider name for logging/debugging */
  readonly name: string

  /**
   * Send a chat request with streaming response.
   * @param messages - Conversation history (system + user + assistant)
   * @param config - LLM configuration (model, temperature, etc.)
   * @returns A ReadableStream that yields text chunks as they arrive
   */
  chat(
    messages: ChatMessage[],
    config: LLMConfig
  ): Promise<ReadableStream<string>>

  /**
   * Check if the provider configuration is valid (API key present, etc.)
   */
  isConfigured(config: LLMConfig): boolean
}

File: chatbot/backend-python/src/llm/interface.py
"""
LLM Provider interface (abstract base class).

Each provider implements chat streaming and returns an async generator
that yields text chunks as they arrive from the API.
"""

from abc import ABC, abstractmethod
from typing import AsyncGenerator, List
from dataclasses import dataclass


@dataclass
class ChatMessage:
    """A single message in the conversation."""
    role: str  # "user", "assistant", "system"
    content: str


class LLMProvider(ABC):
    """Abstract base class for all LLM providers."""

    @property
    @abstractmethod
    def name(self) -> str:
        """Provider name for logging and display."""
        ...

    @abstractmethod
    def is_configured(self) -> bool:
        """Check if the provider has valid configuration (API key, etc.)."""
        ...

    @abstractmethod
    async def chat(
        self,
        messages: List[ChatMessage],
    ) -> AsyncGenerator[str, None]:
        """
        Send a chat request with streaming response.

        Args:
            messages: Conversation history (system + user + assistant).

        Yields:
            Text chunks as they arrive from the API.
        """
        ...
